### PR TITLE
CA-66 Host-Based Routing

### PR DESCRIPTION
**Pull request by franco-caylent**
Implemented host based routing using a map for any applications we want by creating a map of objects like 
```
 "tamr" = {
      length = 1
      instance_ids = [i-0123456789]
      hosts        = ["tamr.*.*"]
      port         = 9100
    }
```

Changes:
- Instantiated EMR cluster with all three applications
- Tested with multiple masters in EMR
- Refactored ALB module for more flexibility and to workaround module limitations
- Renamed variables from dms to host_routing (enable_dms, etc)

### CODE REVIEW

File: modules/alb/variables.tf [hunk 1]
```diff
@@ -48,13 +81,7 @@ variable "tamr_dms_port" {
   default     = "9155"
 }
 
-variable "tamr_dms_hosts" {
```
[the remark this answers is not in the document]
Comment (franco-caylent) [reply modules/alb/variables.tf:51]: Done

File: examples/complete/emr-cluster.tf [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  sg_name_prefix      = format("%s-%s", "example-complete", "-core")
+  egress_protocol     = "all"
+  ingress_protocol    = "tcp"
+}
+
+module "aws-emr-sg-service-access" {
+  source              = "git::git@github.com:Datatamer/terraform-aws-security-groups.git?ref=1.0.0"
+  vpc_id              = module.tamr_networking.vpc_id
+  ingress_cidr_blocks = var.ingress_cidr_blocks
+  egress_cidr_blocks  = var.egress_cidr_blocks
```
[the remark this answers is not in the document]
Comment (franco-caylent) [reply examples/complete/emr-cluster.tf]: Removed

File: examples/complete/emr-cluster.tf [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+  master_instance_fleet_name    = format("%s-%s", "example-complete", "-MasterInstanceFleet")
+  core_instance_fleet_name      = format("%s-%s", "example-complete", "-CoreInstanceFleet")
+  emr_managed_master_sg_name    = format("%s-%s", "example-complete", "-EMR-Spark-Master")
+  emr_managed_core_sg_name      = format("%s-%s", "example-complete", "-EMR-Spark-Core")
+  emr_service_access_sg_name    = format("%s-%s", "example-complete", "-EMR-Spark-Service-Access")
+
+  # Scale
+  master_instance_on_demand_count = 3
+  core_instance_on_demand_count   = 1
+  master_instance_type            = "m4.large"
```
[the remark this answers is not in the document]
Comment (franco-caylent) [reply examples/complete/emr-cluster.tf]: Done

File: modules/alb/main.tf [hunk 4]
```diff
@@ -1,86 +1,143 @@
+
+data "aws_instances" "masters" {
+  filter {
+    name   = "tag:aws:elasticmapreduce:job-flow-id"
+    values = [var.emr_cluster_id]
```
[the remark this answers is not in the document]
Comment (franco-caylent) [reply modules/alb/main.tf:5]: They'll get the following message.
```
 Error: Your query returned no results. Please change your search criteria and try again.
 ```
 We can do a validation for that variable and output an error message if you'd like

File: examples/complete/variables.tf [hunk 5]
```diff
@@ -26,24 +30,24 @@ variable "tamr_dms_port" {
   default     = "9155"
 }
 
-variable "enable_dms" {
+variable "enable_host_routing" {
   type        = bool
-  description = "Enabled the DMS proxying on the port specified in tamr_dms_port"
+  description = "Enable the proxying for EMR, Tamr and DMS"
   default     = true
 }
```
Comment (souza-dan) [inline examples/complete/variables.tf:37]: Do we need this kind of configurability for the examples?

I think we could be more opinionated here and say that this example does set up host based routing and just set that to true within the terraform.

File: modules/alb/main.tf (hunk 4 above)
[the remark this answers is not in the document]
Comment (souza-dan) [reply modules/alb/main.tf:5]: If a user doesn't want to route to EMR, what would they do?

File: modules/alb/main.tf [hunk 6]
```diff
@@ -1,86 +1,162 @@
+
+data "aws_instances" "masters" {
```
Comment (souza-dan) [inline modules/alb/main.tf:2]: I don't think this data object is being used anymore.

Comment (franco-caylent) [reply modules/alb/main.tf:2]: you are right, it was leftover from tests

**Review by souza-dan — APPROVED** [review]
LGTM! Thanks for all the detailed comments!

**Pull request merged**